File: android/src/main/java/org/maplibre/reactnative/components/images/MLRNImagesManager.java
package org.maplibre.reactnative.components.images;

import android.graphics.drawable.BitmapDrawable;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableMapKeySetIterator;
import com.facebook.react.bridge.ReadableType;
import com.facebook.react.common.MapBuilder;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;
import org.maplibre.reactnative.components.AbstractEventEmitter;
import org.maplibre.reactnative.components.AbstractMapFeature;
import org.maplibre.reactnative.components.styles.sources.MLRNShapeSource;
import org.maplibre.reactnative.events.constants.EventKeys;
import org.maplibre.reactnative.utils.ImageEntry;
import org.maplibre.reactnative.utils.ResourceUtils;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;


public class MLRNImagesManager extends AbstractEventEmitter<MLRNImages> {
    public static final String REACT_CLASS = "MLRNImages";

    private ReactApplicationContext mContext;

    @Override
    public String getName() {
        return "MLRNImages";
    }


    public MLRNImagesManager(ReactApplicationContext context) {
        super(context);
        mContext = context;
    }

    @Override
    public MLRNImages createViewInstance(ThemedReactContext context) {
        return new MLRNImages(context, this);
    }

    @ReactProp(name = "id")
    public void setId(MLRNImages source, String id) {
        source.setID(id);
    }

    @ReactProp(name = "images")
    public void setImages(MLRNImages images, ReadableMap map) {
        List<Map.Entry<String, ImageEntry>> imagesList = new ArrayList<>();

        ReadableMapKeySetIterator iterator = map.keySetIterator();
        while (iterator.hasNextKey()) {
            String imageName = iterator.nextKey();
            ImageEntry imageEntry = null;
            if (map.getType(imageName) == ReadableType.Map) {
                ReadableMap imageMap = map.getMap(imageName);
                String uri = imageMap.getString("uri");
                boolean hasScale = imageMap.hasKey("scale") && imageMap.getType("scale") == ReadableType.Number;
                double scale = hasScale ? imageMap.getDouble("scale") : ImageEntry.defaultScale;
                imageEntry = new ImageEntry(uri, scale);
            } else {
                imageEntry = new ImageEntry(map.getString(imageName));
            }
            imagesList.add(new AbstractMap.SimpleEntry<String, ImageEntry>(imageName, imageEntry));
        }

        images.setImages(imagesList);
    }

    @ReactProp(name = "hasOnImageMissing")
    public void setHasOnImageMissing(MLRNImages images, Boolean value) {
        images.setHasOnImageMissing(value);
    }

    @ReactProp(name = "nativeImages")
    public void setNativeImages(MLRNImages images, ReadableArray arr) {
        List<Map.Entry<String, BitmapDrawable>> resources = new ArrayList<>();

        for (int i = 0; i < arr.size(); i++) {
            String resourceName = arr.getString(i);
            BitmapDrawable drawable = (BitmapDrawable) ResourceUtils.getDrawableByName(mContext, resourceName);

            if (drawable != null) {
                resources.add(new AbstractMap.SimpleEntry<String, BitmapDrawable>(resourceName, drawable));
            }
        }

        images.setNativeImages(resources);
    }

    @Override
    public Map<String, String> customEvents() {
        return MapBuilder.<String, String>builder()
                .put(EventKeys.IMAGES_MISSING, "onImageMissing")
                .build();
    }
}

File: android/src/main/java/org/maplibre/reactnative/components/images/MLRNImagesManager.kt
package org.maplibre.reactnative.components.images

import com.facebook.react.bridge.Dynamic
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableType
import com.facebook.react.module.annotations.ReactModule
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager
import com.facebook.react.uimanager.ViewManagerDelegate
import com.facebook.react.uimanager.annotations.ReactProp
import com.facebook.react.viewmanagers.MLRNImagesManagerDelegate
import com.facebook.react.viewmanagers.MLRNImagesManagerInterface
import org.maplibre.reactnative.utils.ImageEntry

@ReactModule(name = MLRNImagesManager.REACT_CLASS)
class MLRNImagesManager(
    private val context: ReactApplicationContext,
) : ViewGroupManager<MLRNImages>(),
    MLRNImagesManagerInterface<MLRNImages> {
    companion object {
        const val REACT_CLASS = "MLRNImages"
    }

    private val delegate: MLRNImagesManagerDelegate<MLRNImages, MLRNImagesManager> =
        MLRNImagesManagerDelegate(this)

    override fun getDelegate(): ViewManagerDelegate<MLRNImages> = delegate

    override fun getName(): String = REACT_CLASS

    override fun createViewInstance(context: ThemedReactContext): MLRNImages = MLRNImages(context)

    @ReactProp(name = "images")
    override fun setImages(
        view: MLRNImages,
        value: Dynamic,
    ) {
        val readableMap = value.asMap()

        if (readableMap == null) {
            view.setImages(emptyList(), context)
            return
        }

        val imagesList = mutableListOf<Map.Entry<String, ImageEntry>>()
        val iterator = readableMap.keySetIterator()

        while (iterator.hasNextKey()) {
            val imageName = iterator.nextKey()
            val imageEntry: ImageEntry =
                when (readableMap.getType(imageName)) {
                    ReadableType.Map -> {
                        val imageMap = readableMap.getMap(imageName)!!

                        val uri = imageMap.getString("uri")
                        val scale =
                            if (imageMap.hasKey("scale")) imageMap.getDouble("scale") else null
                        val sdf = imageMap.hasKey("sdf") && imageMap.getBoolean("sdf")

                        ImageEntry(uri, scale, sdf)
                    }

                    else -> {
                        continue
                    }
                }

            imagesList.add(java.util.AbstractMap.SimpleEntry(imageName, imageEntry))
        }

        view.setImages(imagesList, context)
    }

    @ReactProp(name = "hasOnImageMissing")
    override fun setHasOnImageMissing(
        images: MLRNImages,
        value: Boolean,
    ) {
        images.setHasOnImageMissing(value)
    }
}
